sockeye/inference.py

```diff
@@ -630,6 +635,62 @@ def make_input(sentence_id: int, sentence: str) -> TranslatorInput:
+        Splits oversized sentneces to sentence chunks of size less than max_input_length.
+
+        :param trans_input: List of TranslatorInputs as returned by make_input().
+        :return: List of translation results.
+        """
+        # TODO handle empty inputs, preferably outside inference.py?
+        # TODO avoid accepting and creating lists, use generators?
+
+        # keep input ids to be able to concatenate translated sentence chunks back
+        chunk_markers = []
+        input_chunks = []
+
+        # prepare chunks, keeping track of splitted inputs
+        for trans_input in trans_inputs:
+            if len(trans_input.tokens) > self.max_input_length:
+                logger.debug("Input %d has length (%d) that exceeds max input length (%d). Splitting into chunks of size %d.",
+                             trans_input.id, len(trans_input.tokens), self.buckets_source[-1], self.max_input_length)
+                token_chunks = list(utils.chunks(trans_input.tokens, self.max_input_length))
+                input_chunks.extend(token_chunks)
+                chunk_markers.extend([trans_input.id] * len(token_chunks))
+            else:
+                input_chunks.append(trans_input.tokens)
+                chunk_markers.append(trans_input.id)
+
+        # start translation in batches
+        translations = []  # type: List[Translation]
+        for i in range(0, len(input_chunks), self.batch_size):
+            logger.debug("Translating batch %d", i)
+            batch = input_chunks[i:i + self.batch_size]
+            # underfilled batch will be filled to a full batch size with empty inputs
+            rest = self.batch_size - len(batch)
+            if rest > 0:
+                logger.debug("Extending the last batch to the full batch size (%d)", self.batch_size)
+                batch = batch + [batch[0]] * rest
+            # translate
+            translated_batch = self.translate_nd(*self._get_inference_input(batch))
+            # truncate to remove duplicated translations
+            if rest > 0:
+                translated_batch = translated_batch[:-rest]
+            translations.extend(translated_batch)
+
+        # concatenate translations corresponding to spans of chunk markers with the same input id
+        results = []  # type: List[TranslatorOutput]
+        start = 0
+        while start < len(chunk_markers):
+            _, translations_to_concat = zip(*itertools.takewhile(                                               # while..
+                                            lambda tup: chunk_markers[start] == chunk_markers[start + tup[0]],  # ..it's the same marker (sent id),
+                                            enumerate(translations[start:])))                                   # take pairs of indices & translations.
+            results.append(self._make_result(trans_inputs[len(results)],
+                           self._concat_translations(translations_to_concat)))
```

Comment:
we should avoid the additional overhead of _concat_translations when there is only one translation:
```
results.append(self._make_result(trans_inputs[len(results)],
                          translations_to_concat[0] if len(translations_to_concat) == 1 else self._concat_translations(translations_to_concat)))
```